### ISSUE TITLE
Feature idea: Command swapping

### ISSUE THREAD

**Issue by clemera** [oantolin/embark#28]
Sometimes when you are in a command which uses a minibuffer you want to switch to a related command but keep the current input. For example from `switch-to-buffer` to `find-file` or `project-find-file` and so on. One could define swapping commands for embark classes and then swap commands using an `embak-swap-command`. A similar feature request exists for [ivy](https://github.com/abo-abo/swiper/issues/606). What do you think?

**Comment by oantolin** [comment]
> I think it's going to take some more design work before this is really smooth to use for the `file+buffer` case because different commands want file paths in different formats. For example buffers are usually named after just the file name without the directory, `find-file` wants a path and abbreviates the home directory to `~/`, `project-find-file` also wants paths but does not abbreviate the home directory, etc.

I think 4aa9b61 takes care of this without needing an extra table of input conversion functions.

**Comment by clemera** [comment]
Thank you, great work! Here are my notes:

- `recentf-find-file` is not a command for me (Emacs 26.3).
- I would be interested to have an `embark-input-getters` variable, having it would feel more inline to me with the other hook variables which one can use to adjust the contextual getters but I can live without it of course.
- `describe-symbol` is missing in  `embark-become-help-map`
- For `embark-become-help-map` I would personally rebind keys to have single keystrokes so I can see all the available commands with which-key right away:

```elisp
(defvar embark-become-help-map
  (embark-keymap
   '(("V" . apropos-variable)
     ("O" . apropos-user-option)
     ("C" . apropos-command)
     ("v" . describe-variable)
     ("f" . describe-function)
     ("s" . describe-symbol)
     ("F" . describe-face)
     ("p" . describe-package)
     ("i" . describe-input-method))
   embark-meta-map))
```

**Comment by oantolin** [comment]
Thanks for the feedback! I fixed the typo in the recentf command and switched to your keymap for the help commands (not only does abolishing prefixes help which-key, it's also consistent with all the other keymaps!).

I have no problem with using an `embark-input-getters` list instead. The reason I didn't do it is that it seemed to me that, while different users might want different functions to get the input, probably each user wants only one function for this purpose. It's not like targets or candidates, where there can be many sources. Although, maybe some people switch between Selectrum and Icomplete and want both configured at the same time? I guess if people make fun of me for making this a hook when a function sufficed I can always I claim I did it for consistency. :)

**Comment by oantolin** [comment]
Oh my god! I just realized we've come full circle! The very first embryonic version of embark which was a [reddit comment to you](https://www.reddit.com/r/emacs/comments/gixh12/quickly_grepping_project_in_vanilla_fashion/fqluxs3), was basically `embark-become` (but without the keymaps)!

**Comment by oantolin** [comment]
- Are there any other good circles of mutual becoming we can add?

**Comment by oantolin** [comment]
I added `find-library` to the file+buffer circle. I just realized I very often start `find-file` only to realize I'm far directorywise from the file I want to open but I know the file I want is on the `load-path`.

**Comment by clemera** [comment]
> I guess if people make fun of me for making this a hook when a function sufficed I can always I claim I did it for consistency. :)

:D you can also just blame me for that. I like the hook variable approach because I use a different `completing-read-function` sometimes (I also have a bunch of macros like `with-helm`, `with-ivy`, `with-default` and such to make that easy) and I like that embark then automatically works for all cases it supports.

> Oh my god! I just realized we've come full circle! The very first embryonic version of embark which was a reddit comment to you, was basically embark-become (but without the keymaps)!

Hehe :) Great features often start with reddit comments :)

**Comment by oantolin** [comment]
> Hehe :) Great features often start with reddit comments :)

I'm glad you saw that anyway: I deleted that comment here at GitHub because it was a little wrong: the version in that reddit comment used the top completion, not the un-completed input string.

**Comment by clemera** [comment]
> Are there any other good circles of mutual becoming we can add?

I don't can thing of any useful enough built-ins to add. `occur`, `keep-lines` and `flush-lines` maybe but I don't expect to switch between those. But one could then add `swiper`, `helm-rg` and others to this map which would be nice.

**Comment by oantolin** [comment]
> `occur`, `keep-lines` and `flush-lines` maybe but I don't expect to switch between those. But one could then add `swiper`, `helm-rg` or similar ones to this map which would be nice.

I think that's a good idea. Being a natural place to add swiper or helm-rg is a good reason, but even without that, I think it's nice to be able to switch between keep-lines and flush-lines. (I'd find it useful, "wait, no, I wanted the exact opposite!", happens to me more than I care to admit).

**Comment by oantolin** [comment]
I'm not sure `embark-become-regexp-map` is the best name, but I added a keymap with those 3.

**Comment by clemera** [comment]
Nice, another possible name I can think of would be `embark-become-match-map` (for commands which filter according to matching lines).

**Comment by oantolin** [comment]
I went with your suggestion, I like it a little better than regexp. 883df31

Oh, and I updated `embark--metadata`. I didn't use `embark-input-getters` since I think the string you are actually supposed to pass to `completion-metadata` is the portion of the input string before point. But anyway, the new definition seems like it should be equally appropriate for Selectrum or any completion framework that does actually use the minibuffer.  27e03bd

By the way, I started out a little skeptical about the value of this. I mean, not skeptical that it was worth adding and within the scope of embark, but skeptical that I personally would use it often, but I actually do use `find-file` => `find-library` and `switch-to-buffer` => `find-file` fairly often. Thanks!

Well, it looks like it might be time to close this issue. Is there anything remaining to discuss or implement?

**Comment by clemera** [comment]
> Oh, and I updated embark--metadata. I didn't use embark-input-getters since I think the string you are actually supposed to pass to completion-metadata is the portion of the input string before point. But anyway, the new definition seems like it should be equally appropriate for Selectrum or any completion framework that does actually use the minibuffer. 27e03bd

Thanks, this works BTW I don't know much about fields why did you use `field-beginning`? 

> By the way, I started out a little skeptical about the value of this whole feature. I mean, not skeptical that it was worth adding and within the scope of embark, but skeptical that I personally would use it much, but I actually do use find-file => find-library and switch-to-buffer => find-file fairly often. Thanks!

Glad that you have a use case, too! And thanks for implementing it!

> Well, it looks like it might be time to close this issue. Is there anything remaining to discuss or implement?

No, I don't think so. I have another idea but that is unrelated to this feature: One thing I do not like about `which-key` is that it doesn't give you a buffer I can search and navigate. I experimented a bit and managed to get a hacky alternate version of `embark-keymap-help` which allows me to switch to the help buffer and and search/navigate it and finally select an action with `RET`. The implementation ocurrently only works for `embark-act`. It would be easier and less hacky if we wouldn't use `set-transient-map` and instead read the action key manually using `read-key`. What do you think?

**Comment by oantolin** [comment]
An improved `embark-keymap-help` would be great. The existing one isn't something I particularly like, it's just the quickest thing to put together I could think of.

I am reluctant to use `read-key` instead of `set-transient-map`. For one thing, I bet you meant `read-key-sequence`, instead. :) But seriously, using `set-transient-map` gets you all the [services of the command loop](https://emacs.stackexchange.com/a/55076/2221), some of which are more important than others, but I'd hate to have to reimplement have of them.

**Comment by clemera** [comment]
Thanks for the link, your are right reimplementing the command loop is an issue. I will continue to experiment and if I come up with something better I will open an issue or PR and we can discuss this further.

**Comment by clemera** [comment]
Here is my temporary hack in case you are interested:

```elisp
(defvar embark--help-cmd+ nil)

(add-hook 'embark-pre-action-hook
          (defun embark-which-key-tear-down ()
            ;; get rid of which key window
            (kill-local-variable 'which-key-persistent-popup)
            (kill-local-variable 'which-key-show-prefix)
            (unless which-key-persistent-popup
              (which-key--hide-popup-ignore-command))
            ;; temp hack for presistent help
            (when (and (memq this-command '(which-key-C-h-dispatch embark-keymap-help+))
                       (active-minibuffer-window))
              (embark-keymap-help+)
              (recursive-edit)
              (when (active-minibuffer-window)
                (select-window (active-minibuffer-window)))
              (setq this-command embark--help-cmd+))))

(define-key embark-meta-map
  (kbd "C-h") 'embark-keymap-help+)
(defun embark-keymap-help+ ()
  "Pop up help buffer for current embark keymap."
  (interactive)
  (help-setup-xref (list #'embark-keymap-help+)
                   (called-interactively-p 'interactive))
  ;; fill the buffer
  (with-output-to-temp-buffer (help-buffer)
    (princ
     (substitute-command-keys
      (format "\\{%s}" 'embark--keymap))))
  ;; setup keymaps
  (with-current-buffer (help-buffer)
    (let ((map (make-sparse-keymap)))
      (map-keymap (lambda (ev def)
                    (define-key map (vector ev)
                      (lambda ()
                        (interactive)
                        (setq embark--help-cmd+ def)
                        (exit-recursive-edit))))
                  embark--keymap)
      (set-keymap-parent map (current-local-map))
      (define-key map (kbd "RET")
        (lambda ()
          (interactive)
          (setq embark--help-cmd+ (intern (thing-at-point 'symbol)))
          (exit-recursive-edit)))
      (setq overriding-local-map map)))
  (select-window (get-buffer-window (help-buffer))))
```

**Comment by oantolin** [comment]
This isn't quite what you suggested, but maybe it's good enough: I replaced `embark-keymap-help` with a command `embark-prompt-for-action` that uses `completing-read` to prompt for a bound action (b2c81db). The completion candidates are the actions together with their keybindings, so it also helps learn the bindings. It works with `embark-act`, `embark-act-no-exit` and `embark-become`.

It's not a buffer, but it since it uses `completing-read` you do get nice incremental search. I think it might actually be better than a buffer, since (1) it uses your favorite completion UI, (2) you are unlikely to want to keep the buffer around after choosing an action anyway. (And, if you really do want a buffer showing the actions and bindings, there's always `embark-occur` ---although you get an inert buffer: the candidates don't do anything.)

**Comment by clemera** [comment]
I'm half asleep but this looks pretty cool! I will test in more detail tomorrow, thanks again for your awesome work!

**Comment by clemera** [comment]
This works great! I only needed to make a small adjustment to my `which-key` setup, I also update the snippet for the wiki accordingly.

**Comment by clemera** [comment]
BTW I think using `which-key` as stated in the README does not work, I noticed that when testing this.

**Comment by oantolin** [comment]
Thanks for testing!

It's funny how you do the same thing for prefixes! I do read your blog but I somehow missed that post --it's such a good idea. I'm going to set `prefix-help-command` to a slightly generalized `embark-prompt-for-action` now. :)

By the way, I'm still using `embark-completing-read` as my completion UI and with grid view (which is my default) `embark-prompt-for-action` looks a lot like `which-key`. :)

**Comment by oantolin** [comment]
> BTW I think using `which-key` with default config and the setup from the README does not work, I noticed this while fixing my setup (the one from the wiki).

The setup from the README works with my config and from `emacs -q` (in `emacs -q`, I just loaded embark, loaded which-key, enabled `which-key-mode`, and ran the config in the README). What went wrong when you tried?

**Comment by clemera** [comment]
> By the way, I'm still using embark-completing-read as my completion UI and with grid view (which is my default) embark-prompt-for-action looks a lot like which-key. :)

That sounds nice :) I haven't played much with `embark-completing-read` and live occur yet but I plan to do so (live occur has some problems with selectrum because selectrum own live completion mechanism would have to be turned off during live occur). 

 > What went wrong when you tried?

Which-key did not show up but I did not try it with `emacs -Q`, probably it was something else in my config that caused it, I will check it and report back if it is relevant for `embark`.

What do you think about adding docstrings for `embark-prompt-for-action`? But maybe that isn't great with grid view?

**Comment by oantolin** [comment]
> That sounds nice :) I haven't played much with `embark-completing-read` and live occur yet but I plan to do so (live occur has some problems with selectrum because selectrum own live completion mechanism would have to be turned off during live occur).

Maybe the live occur buffer should have their own hook separate from `embark-occur-mode-hook`. Maybe even liveness should be a minor mode for `embark-occur-mode`.

> What do you think about adding docstrings for `embark-prompt-for-action`?

Well, they would probably be unhelpful for default completion, the `*Completions*` buffer, icomplete, and embark occur grid view. On the other hand they would problably be good for icomplete-vertical, Selectrum and embark occur list view. I'm not sure what the best way to handle this is.

I wanted at some point to add them as embark annotations but didn't do it because I realized that would leave icomplete-vertical and Selectrum in the lurch. Selectrum also has an annotation feature, right? Maybe they could be added as annotation's there too? (But that probably involves writing a new command that uses Selectrum's read function instead of `completing-read`, right?)

There also the issue of whether they are added as part of the candidate text or only in apperance (in an overlay or text property), that is, would one want to match against the docstring text too or not?

Maybe, since I think I definitely don't want docstrings for default completion, the `*Completions*` buffer, icomplete, or embark occur grid view, the best thing to do is to think about adding docstrings separately for each vertical UI. For embark occur list view and for Selectrum probably the best option is to add them as annotations.

**Comment by clemera** [comment]
> (But that probably involves writing a new command that uses Selectrum's read function instead of completing-read, right?)

Selectrum will soon handle the `annotation-function` property of completion tables. I thought that would be the best way to add them. Why wouldn't you like them for the `*Completions*` buffer? I don't know how `icomplete` handles annotations. One could also add them only for selectrum by propertizing the strings with `selectrum-candidate-display-suffix` or `selectrum-candidate-display-right-margin`, you don't have to use `selectrum-read`. 

> There also the issue of whether they are added as part of the candidate text or only in apperance (in an overlay or text property), that is, would one want to match against the docstring text too or not?

With annotations added by `annotation-function` meta data they wouldn't be searchable from within the minibuffer but one can switch to the `*Completions*` buffer and search them. Maybe `embark-occur` should also show annotations so they would be searchable from there?

**Comment by oantolin** [comment]
> Why wouldn't you like them for the `*Completions*` buffer?

I have not looked in depth at the `annotation-function` metadata property, so I could be wrong about any this, but here's what I believe about it:

First, in the built-in stuff it is only used in the `*Completions*` buffer (in particular Icomplete doesn't show them). The `*Completions*` buffer always has at least 2 columns (which is why for live-completions I wrote a single-column view), so there's never much space. So `annotation-function` annotations tend to be small things that can fit into a packed grid view. 

I only know of 4 examples of annotations in Emacs: 

1. keybindings in M-x, 

2. for `C-x 8 RET` the candidates are unicode character **names**, the annotation is the actual character, 

3. there's a `<f>` or `<v>` to distinguish functions from variables somewhere (probably `describe-symbol`?), 

4. Unix signals in `proced` have a description as an annotation.

Of these only the last are long, and they look terrible. :)

![Screenshot 2020-06-15 at 4 22 17 AM](https://user-images.githubusercontent.com/2635478/84641544-c6c56300-aec0-11ea-8347-af09ce6bd84b.png)

> Maybe embark-occur could also show annotations (for the list view) if they are present so they would be searchable from there and for the grid view they could be omitted.

I think, if anything, maybe embark-occur should show `annotation-function` annotations in grid view, and show the embark annotations in list view (which it already does). Or maybe it should concatenate the `annotation-function` annotations with the embark annotations and show them in list view. I'm not sure.

> Selectrum will soon handle the annotation-function meta data of completion tables.

Unless there are more example I'm unaware of, I am not sure it's really worth it. Doesn't selectrum already show keybindings (and even docstrings) in M-x? Well, it's easy to implement so maybe it is worth it even just to see SNOWMAN WITHOUT SNOW in `C-x 8 RET`.

> With annotations added by annotation-function meta data they wouldn't be searchable from within the minibuffer but you can switch to the `*Completions*` buffer and search them from there.

I do like the idea of showing them, not having them searchable from the minibuffer, but being able to search in them by switching to the `*Completions*` buffer (or an embark occur buffer).

So, to summarize: putting first line of docstrings in `annotation-function` is completely doable and we can augment embark and selectrum to use them, but `*Completions*` will look awful. :)

**Comment by oantolin** [comment]
This choice can be revisited but for now I've implemented using `annotation-function` as a fallback if `embark-annotator-alist` is slient on what to do. This means `annotation-function` annotations will be displayed in embark occur list view if embark doesn't have it's own annotation for a given context. 4243d6b
 
![Screenshot 2020-06-15 at 4 55 39 AM](https://user-images.githubusercontent.com/2635478/84644422-c16a1780-aec4-11ea-857c-a9017b779df7.png)

**Comment by oantolin** [comment]
I removed the parenthesis (or angle brackets) if `annotation-function` uses them. By the way, beware of `completion-extra-properties`, it too can carry the `annotation-function` when the `completion-metadata` doesn't.

**Comment by clemera** [comment]
> This choice can be revisited but for now I've implemented using annotation-function as a fallback if embark-annotator-alist is slient on what to do.

I like that! 

> By the way, beware of completion-extra-properties, it too can carry the annotation-function when the completion-metadata doesn't.

Thanks for that! I thought those are only relevant for `completion-in-region`. 

> But I tend to use the default frame size which is about 80 characters wide. It seems unwise to have annotations look messy at the default frame size.

I always thought everyone uses a maximized frame :)

**Comment by oantolin** [comment]
Oh, I forgot, but I will add the docstrings as annotations in `embark-prompt-for-action`. By the way, that function doesn't just prompts for actions, it's also used by `embark-become`'s `C-h`. What's a good name to cover both cases?

>> By the way, beware of completion-extra-properties, it too can carry the annotation-function when the completion-metadata doesn't.

> Thanks for that! I thought this would only be relevant for `completion-in-region`. Just updated my PR for selectrum with that.

I should have said explicitly that the Unix signals in `proced-send-signal` are an example of the use of `completion-extra-properties`.

> Currently not, you have probably seen that with `counsel-M-x`.

Huh. Yeah, probably, I did install Counsel a week or two ago to test something, on the same day I tested various things about Selectrum too, so I could easily have gotten mixed up.

> Another lesson for me to avoid making to much assumptions about how other people use Emacs :)

Well, particularly defaults: for any given default some people are bound to use it just because of inertia. I do maximize Emacs for certain things, basically to annotate PDFs with `pdf-tools` or to write in LaTeX --which I do with Emacs maximized split in two windows, one with the source and one with the compiled PDF. For other things I typically keep the Emacs frame as it is when it opens.

**Comment by oantolin** [comment]
OK, now I've added the docstring in an annotation-function and learned something else: the annotations need a leading space because `*Completions*` just concatenates them to the candidates! In this screenshot you can see both an embark occur list view and a `*Completions*` buffer (in a maximized frame). :(

![Screenshot 2020-06-15 at 10 23 31 AM](https://user-images.githubusercontent.com/2635478/84676109-b332f000-aef2-11ea-8332-77238ed78256.png)

**Comment by clemera** [comment]
> What's a good name to cover both cases?

Maybe `embark-help`?

 > I should have said explicitly that the Unix signals in proced-send-signal are an example of the use of completion-extra-properties.

Thanks, I discovered that when looking at `proced`. It's kind of confusing to me because the `:exit-function` of `completion-extra-properties` is only used with `completion-in-region` and the docs don't make that clear.

> For other things I typically keep the Emacs frame as it is when it opens.

I'm curious: As a floating window or do you use a tiling WM?

**Comment by clemera** [comment]
> In this screenshot you can see both an embark occur list view and a *Completions* buffer (in a maximized frame). :(

Personally I don't use the *Completions* buffer but maybe showing the annotations could be an embark user option.

**Comment by oantolin** [comment]
> Maybe `embark-help`?

Or even more specifically, `embark-keymap-help`, which is the name of the old function this replaces! :D
 
> Thanks, I discovered that when looking at `proced`. It's kind of confusing to me because the `:exit-function` of `completion-extra-properties` is only used with `completion-in-region` and the docs don't make the distinction how those properties are used clear.

I guess, as often happens, the source is the true documentation...

> I'm curious: As a floating window or do you use a tiling WM?

I used i3 when I had a personal Linux laptop. At the office (which I haven't been to in months, due to the pandemic) I use some old version of Ubuntu I'm not allowed to change, so I just stick to either Unity or Gnome 3 (I forget which one it was, actually). At home I have a Chromebook and just let ChromeOS  manage windows however it does by default --so, floating windows, but easy snapping to the left or right half, which is enough tiling for me.

> Maybe truncating the lines could improve the display, personally I don't use the `*Completions*` buffer.

I don't use it either, not since embark-occur!

> Maybe the user should be able to control the display via an embark user option.

I think a user option is probably reasonable, and it can default to having full docstrings, you know, for us non-`*Completions*` users. How about something like `embark-keymap-help-docstrings` with values:

- `nil`: no docstrings.
- `t`: full docstrings.
- `truncate`: truncate docstrings so that `key -> cmd docstring` is at most half the frame-width.

**Comment by clemera** [comment]
> I guess, as often happens, the source is the true documentation...

I have to admit I never really looked at minibuffer.el in detail.

> I used i3 when I had a personal Linux laptop. At the office (which I haven't been to in months, due to the pandemic) I use some old version of Ubuntu I'm not allowed to change, so I just stick to either Unity or Gnome 3 (I forget which one it was, actually). At home I have a Chromebook and just let ChromeOS manage windows however it does by default --so, floating windows, but easy snapping to the left or right half, which is enough tiling for me.

Interesting, thanks. I'm on Gnome 3 currently. There are some tiling extensions for it but I haven't tried them because I don't have a need for much window management. Most of the time I have two windows, one per monitor: Firefox and Emacs :)

> I think a user option is probably reasonable, and it can default to having full docstrings, you know, for us non-*Completions* users. How about something like embark-keymap-help-docstrings with values:
>    nil: no docstrings.
    t: full docstrings.
    truncate: truncate docstrings so that key -> cmd docstring is at most half the frame-width.

Sounds good :+1:

**Comment by oantolin** [comment]
> Interesting, thanks. I'm on Gnome 3 currently. There are some tiling extensions for it but I haven't tried them because I don't have a need for much window management. Most of the time I have two windows, one per monitor: Firefox and Emacs :)

Me too, I don't care about window management anymore: the browser can manage it's tabs well enough, and Emacs can manage its windows well-enough. I occasionally open a second Emacs frame, or put the browser and Emacs side by side, but that's the fanciest window management I ever do. :)

OK, I implement the user option. Here's what setting it to `truncate` looks like in the `*Completions*` buffer (I think only `*Completions*` users would ever want `truncate`):

![Screenshot 2020-06-15 at 5 56 55 PM](https://user-images.githubusercontent.com/2635478/84713991-fb253780-af31-11ea-86dc-2fdb262a9163.png)
![Screenshot 2020-06-15 at 5 56 35 PM](https://user-images.githubusercontent.com/2635478/84713995-fbbdce00-af31-11ea-827f-264739cda57c.png)

**Comment by clemera** [comment]
Works great, thanks a lot for implementing! There is one last thing I noticed with annotations I have opened a #30 for that.
